### ISSUE TITLE
SQLite syntax error when SQLRequest involves string interpolation

### ISSUE THREAD

**Issue by hansengj** [groue/GRDB.swift#1332]
### What did you do?

Executed "try request.fecthAll(db) where request involved string interplation
where
let table = "places"
let request = SQLRequest<Row>("SELECT json FROM \(table)")

### What did you expect to happen?

No SQLite syntax error. SQLRequest documentation illustrates use of string interpolation in a request.

### What happened instead?

Got error:
SQLite error 1: near "?": syntax error - while executing `SELECT json FROM ?`

### Environment

**GRDB flavor(s):** GRDB
**GRDB version:** v6.6.1
**Installation method:** manual
**Xcode version:** 14.2
**Swift version:** 5.7.2
**Platform(s) running GRDB:** iOS 
**macOS version running Xcode:** 13.2 on macMini M1

### Demo Project
Code:
let table = "places"
let request = SQLRequest<Row>("SELECT json FROM \(table)")
do {
       try dbReader.read { db in
               let rows = try request.fetchAll(db)  //try Row.fetchAll(db, request) also fails
               ....
        }
 } catch {
        print("\(error)")
}

Request: print("Select request: \(request)")

Select request: SQLRequest<Row>(adapter: nil, sqlLiteral: GRDB.SQL(elements: [GRDB.SQL.Element.sql("SELECT compressedJSON FROM ", []), GRDB.SQL.Element.expression(GRDB.SQLExpression(impl: GRDB.SQLExpression.(unknown context at $1044ffcf4).Impl.databaseValue("places")))]), cache: nil)

Note unknown context for "places".

Swift 5 may be suspect since it introduced a new version of string interpolation.

No syntax error if replace \(table) with actual string "places", i.e., fetchAll executes properly.

**Comment by groue** [comment]
Hello @hansengj,

SQLite does not support providing table names as arguments:

```swift
// Invalid SQLite syntax: SELECT * FROM ?
let table = "player"
SQLRequest("SELECT * FROM \(table)")
```

Such interpolation can only be used for values:

```swift
// OK: SELECT * FROM player WHERE name = ?
let name = "O'Brien"
SQLRequest("SELECT * FROM player WHERE name = \(name)")
```

You can provide your table name as *raw sql*:

```swift
// OK: SELECT * FROM player
let table = "player"
SQLRequest("SELECT * FROM \(sql: table)")
```

Note that when you inject raw SQL into your SQL requests, it is your responsibility to avoid SQL injection.

**Comment by hansengj** [comment]
Thank you for the workaround. In GRDB's documentation, you may want to mention this fact, it it doesn't already (e.g., as an example in SQLRequest's documentation you see in Xcode).

**Comment by groue** [comment]
I'm not sure that the [SQL Interpolation](https://github.com/groue/GRDB.swift/blob/master/Documentation/SQLInterpolation.md) guide needs to be updated. It uses the words "value" and "values" when appropriate. I agree that not everybody knows that table names are not "values" according to SQLite. But many things are not (operators, keywords, etc). I won't duplicate or even extend the SQLite documentation. I consider that it is OK to answer your question in an issue. I also consider that it is OK that people learn things about SQLite.

**Comment by hansengj** [comment]
Fair enough. The link to SQL Interpolation suffices. I did not know table names are not SQLite values, which was my responsibility. Not being a database guy, I prefer to work at the GRDB level. But since I use raw SQL statements, I need to have some understanding of SQLite. In this case, my knowledge was insufficient. I appreciate your detailed  explanation in the issue response and how to modify my code so it worked!

I understand your stance about updating the guide, but it wouldn't hurt if you added a foot note in SQL Interpolation that states table names are not SQLLite values and how to use them in interpolation, since it probably is a common use case. Not all users of GRDB are experts in SQLite.

It might help if you understood how I use GRDB. All of my app’s session state is saved to the database as JSON - compressed and then converted to a base64EncodedString. Each state related class has handed coded JSON encode() and decode() methods. When class data changes, these methods are updated. My tables are simple: 1 column with a dbID to uniquely access a row and 1 column of JSON. All of my database access methods use raw SQL. I don’t have to change database access methods when data changes in my classes.

By compressing the JSON saved to the database, its size was reduced by 20X. The time to decompress it when extracted from the database is not significant.

Thanks again for you quick responses and explanations!

—Gil

**Comment by groue** [comment]
> I understand your stance about updating the guide, but it wouldn't hurt if you added a foot note in SQL Interpolation that states table names are not SQLLite values and how to use them in interpolation, since it probably is a common use case.

"My use case is probably a common use case" is a frequent bias, but still a bias 😉 As the maintainer of the repo, and as the one who answers most support requests, I can tell you that no, this is not a frequent misuse of the library.

I quite understand your focus on table names, because we all easily focus on our most recent pain point. It's frustrating to spend time debugging such as "stupid issue". Yet your misstep is not the only possible misstep. A documentation that mentions and warns against every possible misinterpretation is not a good documentation. And I'm now sure that you will remember what's an argument and a value for SQLite. Consider yourself vaccinated against this error 🙂 

> Not all users of GRDB are experts in SQLite.

This is absolutely true. Some actually increase their SQLite knowledge with GRDB, and I'm quite proud of it. SQLite is a remarkable piece of software. https://twitter.com/BelleBCooper/status/1534070340850094080

**Comment by hansengj** [comment]
Amen to the Tweet link.

The next time I run into an issue, I will first dig deeper into SQLite for a resolution. Lesson learned.

**Comment by groue** [comment]
I hope I didn't sound rude! In all cases, thank you for opening this issue. It's a pleasure to answer support requests and unlock fellow developers.

**Comment by hansengj** [comment]
We are good. I respect you for sticking to your principals.  —Gil

**Comment by groue** [comment]
> I flag this issue as "needs revisiting", because the answer I wish I could give was:
> 
> ```swift
> // OK: SELECT * FROM player
> let table = Table("player")
> SQLRequest("SELECT * FROM \(table)")
> ```
> 
> Unfortunately, there is no built-in support for `Table` in SQL interpolation yet. This will come in a future version.

Done in 17906dc505334f59e65533807f292716d97500b7

**Comment by hansengj** [comment]
This would be a great solution. I’m glad you “revisited” my issue. Thanks. —Gil

**Comment by groue** [comment]
> ```swift
> // OK: SELECT * FROM player
> let table = Table("player")
> SQLRequest("SELECT * FROM \(table)")
> ```

Interpolation of `Table` has shipped in v6.7.0

**Comment by hansengj** [comment]
Thank you. I like the way you inform people on the status of their issue(s). Very professional.  —Gil
